### DISCUSSION TITLE
Automation in homeapp - timer auto off

### DISCUSSION THREAD

**Discussion by aLAN-LDZ** [HomeSpan/HomeSpan#212] [Q&A] [answered]
Hi

I have a problem with automation, I made a scene in the home app to start the fan and turn it off after 5 minutes. When I test the automation, the fan starts but does not turn off :(

`struct DEV_Fan : Service::Fan {               
  int FanPin = 4;                                       // pin number
  SpanCharacteristic *power;                       
  
  DEV_Fan() : Service::Fan(){       // constructor() method

    power=new Characteristic::Active(0, true);                 
    this->FanPin=FanPin;                            
    pinMode(FanPin,OUTPUT);
    digitalWrite(FanPin,!power->getVal());                          
    
  } // end constructor

  boolean update(){                              // update() method

    digitalWrite(FanPin,!power->getNewVal());      
   
    return(true);                               // return true
  
  } // update
};`

**Comment by HomeSpan** [comment]
Automations and scenes are handled by the Home App and a HomeKit Hub (if used).  There is nothing specific in HAP-R2 or in any HomeKit device that controls automations.  Others have reported similar problems under certain circumstances (both with HomeSpan and also with commercial products).  I just tested my own setup by turning on a HomeSpan-controlled light and then asked Siri to "turn off the light in 10 minutes."  Siri acknowledged and inserted an automation to turn off the light at a time that was t+10 minutes.  After 10 minutes the light went off as expected.

Do you have a Home Hub (such as HomePod)?  I use my HomePods a a hub and disabled the ability of my Apple TV to be a Home Hub since I found HomeKit always preferenced that device, and it always had problems connecting to HomeKit.

**Comment by aLAN-LDZ** [comment]
yes, I have apple tv, unfortunately what I wouldn't do, it won't work :( eh

**Comment by aLAN-LDZ** [comment]
while for other devices, be it ikea or xiaomi, the automation works. it does not work only in the case of homespan, it will be possible to turn on the device but it will not turn off after the set time. I suspect that since it works with other producers, the problem is somewhere on your side

**Comment by HomeSpan** [comment]
maybe something to do with whether using a bridge or not?  Is the Ikea device set up as a bridge?  How about your HomeSpan device?  The HomeSpan device I tested was NOT set up as a bridge.  That said, I tested it with a light. Just to be sure, I'll test this with the fan portion of the device as well.  I'm not creating any scenes - just turning on the device and asking Siri to turn it off in 10 minutes.  Will try now and let you know if it works.

**Comment by HomeSpan** [comment]
I'm glad I tested this for the fan.  It does not work.  To be clear, I have a single HomeSpan device with an Accessory containing two Services.  One is a Fan Service and the other is a LightBulb Service.  If ask Siri to to turn off the light after 2 minutes, it works.  If I ask Siri to turn off the fan after two minutes it does not work.  If I ask Siri to turn off the entire device after 2 minutes, it also does not work.   It seems like others have found some similar issues as well with commercial products (https://www.reddit.com/r/HomeKit/comments/la6sm7/homekit_automation_turn_off_after_not_working/).  I have a Lutron-controlled Fan I will try this with as well.

**Comment by aLAN-LDZ** [comment]
I have the Homespan set to the bridge and I have three accessories in it, today I will check with the switch, maybe it will be a solution to this error. Let me know if the factory fan will work. Thanks for the research.

Update: if we create a switch and set it as a fan, the automation works fine.

**Comment by HomeSpan** [comment, marked as answer]
SOLVED!

Here was the issue:  For integer-based ON/OFF Characteristics, such as Characteristic::Active, the Home App would send a 0 for OFF, and a 1 for ON, which is correct and consistent with HAP-R2.  ***BUT***...when using Home App AUTOMATIONS, the Home App would instead send the words *true* or *false*, even for integer-based Characteristics, which is ***inconsistent*** with HAP-R2.

This is not a problem for boolean-based Characteristics, such as Characteristic::On, since HomeSpan (per HAP-R2) will accept the numbers *0* for off, and *1* for on, as well as the words *false* for off, and *true* for on.

Since Lights and Switches use Characteristic::On, there are no issues with automation.  But Fans use Characteristic::Active, which is defined as a UINT8.  Hence HomeSpan would reject the words *true* and *false*, since they are unexpected for this type of Characteristic.

I'm not sure why the Home App and Siri both uses numbers (0/1), as expected, but the automations routines use words (true/false).  Regardless of the reasons, the fix was easy:  I updated the HomeSpan parser to interpret true/false as 1/0 for all integer-based Characteristics, so it properly handles requests from automation routines that use true/false instead of 1/0.

The fix for this will be in the next patch release, but you can try it out now in the dev branch.

(I wonder if some commercial products suffer from the same issue as well).
